containers.sh

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+images="${PWD}/docker/images"
+
+log(){
+  echo "$(date +'%T') $1"
+}
+
+networks(){
+  docker network create \
+    todo-app
+}
+
+volumes(){
+  docker volume create \
+    todo-mysql-data
+}
+
+todo_app(){
+  docker run \
+    --detach \
+    --publish 3000:3000 \
+    --workdir "/app" \
+    --volume "${images}"/getting-started/app:/app \
+    --network todo-app \
+    --env MYSQL_HOST=mysql \
+    --env MYSQL_USER=root \
+    --env MYSQL_PASSWORD=secret \
+    --env MYSQL_DB=todos \
+    node:12-alpine \
+    sh -c "yarn install && yarn run dev"
+}
+
+todo_db(){
+  docker run \
+    --detach \
+    --network todo-app \
+    --network-alias mysql \
+    --volume todo-mysql-data:/var/lib/mysql \
+    --env MYSQL_ROOT_PASSWORD=secret \
+    --env MYSQL_DATABASE=todos \
+     mysql:5.7
+}
+
+# Create networks for containers
+log "Creating docker networks..."
+OUTPUT=$(networks)
+
+# Create volumes for containers
+log "Creating docker volumes..."
+OUTPUT=$(volumes)
+
+# Spawn todo-list node.js app container
+log "Spawning todo app container..."
+OUTPUT=$(todo_app)
+
+# Spawn todo-list database
+log "Spawning todo database..."
+OUTPUT=$(todo_db)
```

Comment:
When running in a CI pipeline, you don't want the containers to run indefinitely, but rather you want to execute some sort of validation that they work, and then shut them down. With docker-compose it would be something like (pseudocode):
```
docker-compose up
./run-test.sh (checks that containers work)
docker-compose down
```

In this case, you would need to `docker stop` the containers one at a time.

The `run-test.sh` script could just be grepping for output in the container logs, for example, eg. `docker logs <container_name> | grep "todo app is running"`. Another option is using curl: `curl -ILk http://<container>:3000/` and relying on a HTTP 200 for validation.

Reply:
Decided to go with grep for some port and mysqld listeners. 
However, since the containers need a little time launching before I start DDOS'ing them with all my grep commands, I added a sleep timer to at least 'solve' the problem for now.